Initial documentation of modulesync (#134)
projectsyn/documentation: docs/modules/ROOT/pages/explanations/modulesync.adoc

That's, if there are changes in the component.

== Overriding variables

Variables defined in `config_defaults.yml` in the control repository can be overwritten in `.sync.yml` within a component.

[CAUTION]
====
The merge behaviour is different to the one used within the Syn configuration hierarchy.

.Content of `config_defaults`
[source,yaml]
----
File:
  variables:
    feature1: bar
    feature2: baz
----
If a key is overwritten in the module, it will overwrite the whole `variables` key!

.Content of `.sync.yml`
[source,yaml]
----
File:
  variables:
    feature1: customized
----
The `feature2` key would be removed in the final config.
====

== Keep your component up to date

A component created with `commodore component create` will be ready to be managed by `modulesync-control` of Project Syn.
In order for this to happen, you have to add your component to https://github.com/projectsyn/modulesync-control/blob/master/managed_modules.yml[managed_modules.yml].

== Rules established by `modulesync-control`

Components kept in sync by https://github.com/projectsyn/modulesync-control[modulesync-control] must adhere to some rules.

Don't edit the following files:

* `Makefile`
+

> If you need to add new make targets, you can add them to new files (use `.mk` as file ending.

Reviewer: ```suggestion
If you need to add new make targets, you can add them to new custom Makefiles (suggested file ending `.mk`).
```

You can include them in `Makefile` by adding them to the list variable `Makefile.includes`.

* `.gitignore`
+
If you have more things to exclude, add them to `.gitignore.additionalEntries`

> == Make your existing component ready for ModuleSync

Reviewer: This section would probably better fit in a how-to page

* Create `.sync.yml` with the following content:
+
[source,yaml]
----
:global:
  componentName: my-component <1>

docs/antora.yml:
  name: my-component <1>
  title: My Component <2>

LICENSE:
  year: 2021 <3>
----
<1> The component name
<2> A more human friendly component name

> <3> The year the component got created

Reviewer: ```suggestion
<3> The year the component got created (see existing `LICENSE` file)
```

The template repository for Project Syn is located at https://github.com/projectsyn/modulesync-control[modulesync-control].

https://github.com/voxpupuli/modulesync[ModuleSync] allows to have a set of templates in a control repository.
Those templates then can be applied to several other repositories.
Originally, ModuleSync was written for Puppet Modules.
We have found, that this tool also fits our needs for Commodore components.

The templates are located in the folder `moduleroot`.
Each templates takes a set of variables as input which are defined in a file.
In the control repository the variables are defined in `config_defaults.yml`.
In the component they are located in `.sync.yml`.

The managed components are listed in `managed_modules.yml`.
Each entry refers to a Git repository hosted on GitHub and has the form `<namespaces>/<name>`.
If no namespace is given, the default namespace in `modulesync.yml` will be used.

== Applying changes

Changes in the control repository pushed to `master` will trigger the execution of ModuleSync.
A pull request will be created for each managed component.
That is, if there are changes in the component.

== Overriding variables

> Variables define in `config_defaults.yml` in the control repository, can be overwritten in `.sync.yml` within a component.

Reviewer: ```suggestion
Variables defined in `config_defaults.yml` in the control repository can be overwritten in `.sync.yml` within a component.
```

> [CAUTION]
> The merge behaviour is different to the one used within the Syn configuration hierarchy.
> 
> // TODO describe merge behaviour.

Reviewer: ```suggestion
[CAUTION]
====
The merge behaviour is different to the one used within the Syn configuration hierarchy.

.Content of `config_defaults`
[source,yaml]
----
File:
  variables:
    feature1: bar
    feature2: baz
----
If a key is overwritten in the module, it will overwrite the whole `variables` key!

.Content of `.sync.yml`
[source,yaml]
----
File:
  variables:
    feature1: customized
----
The `feature2` key would be removed in the final config.
====
```

= Keep component boilerplate in sync

[abstract]
The creation of a new Commodore component yield several boilerplate files.
Some of those files are meant to be altered by a component author.
Others should be equal throughout all components managed by Project Syn.
As with all code, things will change.
As a result, components will drift apart over time.
To battle this issue, we make use of https://github.com/voxpupuli/modulesync[ModuleSync] to keep that boilerplate in sync.
The template repository for Project Syn is located at https://github.com/projectsyn/modulesync-control[modulesync-control].

https://github.com/voxpupuli/modulesync[ModuleSync] allows to have a set of templates in a control repository.
Those templates then can be applied to several other repositories.
Originally, ModuleSync was written for Puppet Modules.
We've found, that this tool also fits our needs for Commodore components.

> The templates are located in the folder `moduleroot`.

Reviewer 2: ```suggestion
The templates are located in the folder https://github.com/projectsyn/modulesync-control/tree/master/moduleroot[`moduleroot`].
```

Each templates takes a set of variables as input which are defined in a file.

> In the control repository the variables are defined in `config_defaults.yml`.

Reviewer 2: ```suggestion
In the control repository the variables are defined in https://github.com/projectsyn/modulesync-control/blob/master/config_defaults.yml[`config_defaults.yml`].
```

In the component they're located in `.sync.yml`.

> The managed components are listed in `managed_modules.yml`.

Reviewer 2: ```suggestion
The managed components are listed in https://github.com/projectsyn/modulesync-control/blob/master/managed_modules.yml[`managed_modules.yml`].
```

Each entry refers to a Git repository hosted on GitHub and has the form `<namespaces>/<name>`.

> If no namespace is given, the default namespace in `modulesync.yml` will be used.

Reviewer 2: ```suggestion
If no namespace is given, the default namespace in https://github.com/projectsyn/modulesync-control/blob/master/modulesync.yml[`modulesync.yml`] will be used.
```

== Applying changes

Changes in the control repository pushed to `master` will trigger the execution of ModuleSync.
A pull request will be created for each managed component.
That's, if there are changes in the component.

== Overriding variables

Variables defined in `config_defaults.yml` in the control repository can be overwritten in `.sync.yml` within a component.

[CAUTION]
====

> The merge behaviour is different to the one used within the Syn configuration hierarchy.

Reviewer 2: ```suggestion
The merge behaviour is different to the one used within the Project Syn configuration hierarchy.
```

.Content of `config_defaults`
[source,yaml]
----
File:
  variables:
    feature1: bar
    feature2: baz
----
If a key is overwritten in the module, it will overwrite the whole `variables` key!

.Content of `.sync.yml`
[source,yaml]
----
File:
  variables:
    feature1: customized
----
The `feature2` key would be removed in the final config.
====

== Keep your component up to date

> A component created with `commodore component create` will be ready to be managed by `modulesync-control` of Project Syn.

Reviewer 2: How do we keep https://github.com/projectsyn/commodore/tree/master/commodore/component-template in sync with the ModuleSync template?

Reviewer: Manually as of now. The component template is used by cookiecutter to fill in the defaults in case the component is not intended to be managed by modulesync.
Since the directory structure is also different than moduleroot, we can't manage this component template from modulesync either.

Reviewer 2: Which one do we as Project Syn consider the source of truth in the future? Maybe we should document that somehow.

Reviewer: It's also important to know that the files aren't 100% the same. 
for example, when modulesync creates a PR for the first time, some files will receive a comment at the top like "This file is managed by modulesync - do not edit" along with additional information where to customize it. 
When a user creates a new component with `commodore component new`, the fresh component is not managed by modulesync (and might never be) then such comments are confusing if they would be part of the template.

Reviewer: > Which one do we as Project Syn consider the source of truth in the future? Maybe we should document that somehow.

This is a hard one. I'm not sure there can be a single source of truth.

On one hand, the component template could be source of truth since it defines the initial structure of a component, even if that component never gets managed by modulesync (e.g. private component, PoC etc)

On the other hand, the initial component template may be incomplete in its features. for example, there are components that have unit tests - something today's template doesn't provide. Yet the modulesync template may easily add CI/CD workflows to enable unit testing with a single switch.

Generally I tend to use modulesync's template to be the source of truth, since it's easier to make backport the changes to the component template than the other way round. Once a module gets managed by modulesync, the component is subject to its templating engine anyway and the initial template wouldn't really matter anymore.

Reviewer: Do you seen an action point from this discussion?

Reviewer 2: No, I think it's OK that we at least discussed it. I can live with this situation, and we'll see how it will behave in the future when making full use of this new feature.

In order for this to happen, you have to add your component to https://github.com/projectsyn/modulesync-control/blob/master/managed_modules.yml[managed_modules.yml].

== Rules established by `modulesync-control`

Components kept in sync by https://github.com/projectsyn/modulesync-control[modulesync-control] must adhere to some rules.

Don't edit the following files:

> * `Makefile`

Reviewer 2: ```suggestion
`Makefile`::
```

Maybe better use this syntax on this list?

Reviewer: I agree, looks better

+
If you need to add new make targets, you can add them to new custom Makefiles (suggested file ending `.mk`).
You can include them in `Makefile` by adding them to the list variable `Makefile.includes`.

> * `.gitignore`

Author: ```suggestion
`.gitignore`::
```

A component created with `commodore component create` will be ready to be managed by `modulesync-control` of Project Syn.
In order for this to happen, you have to add your component to https://github.com/projectsyn/modulesync-control/blob/master/managed_modules.yml[managed_modules.yml].

== Rules established by `modulesync-control`

Components kept in sync by https://github.com/projectsyn/modulesync-control[modulesync-control] must adhere to some rules.

Do not edit the following files:

* `Makefile`
+
If you need to add new make targets, you can add them to new files (use `.mk` as file ending.
You can include them in `Makefile` by adding them to the list variable `Makefile.includes`.

* `.gitignore`
+
If you have more things to exclude, add them to `.gitignore.additionalEntries`

Further:

> * Ensure `LICENSE.year` is set to the year your component got created.

Reviewer: This will be default in new components with this PR: https://github.com/projectsyn/commodore/pull/335
